### ISSUE TITLE
Theme name privacy

### ISSUE THREAD

**Issue by zsugabubus** [tridactyl/tridactyl#1586]
Currently, a class is set on the root `html` element to indicate active theme, but it can be easily used by a malicious site to (nearly) uniquely identify an user if she uses a custom theme.

On the other hand I think it's very inconvenient to always use `:root.TridactylThemeMyCustomThemeName123` and a bunch of `!important`s (just to override default styles that you want to get rid of anyway) in theme files. In addition, builtin theme style sheets are also always loaded which is also a bad thing (performance drawback, memory footprint). I think this whole issue could be solved by just loading active theme file and no others. Seems to simple, so is there any reason it is not used already?

**Comment by cmcaine** [comment]
There are only a few thousand tridactyl users so presence-of-tridactyl-iframe + just a few bits of entropy will uniquely identify each user anyway, if any tracking js ever learns to look for tridactyl elements.

I don't recall the reason we use the classnames rather than loading and unloading theme files, but it wouldn't provide privacy either if an attacker knew to hash all the loaded stylesheets, for example.

**Comment by bovine3dom** [comment]
I have no recollection as to why we don't just load the CSS required. It could be that it made making small themes that only change a little of Tridactyl's UI a little easier.

The PRs that added this functionality are https://github.com/tridactyl/tridactyl/pull/579 and https://github.com/tridactyl/tridactyl/pull/537. If you can work out why we do it the way we do currently, and if that reason is no longer valid, I'd be happy to accept a PR that changes the behaviour :)

**Comment by zsugabubus** [comment]
Okay, I'm happily dig into the issue, both from theme and privacy sides.

**Comment by zsugabubus** [comment]
My ideas, if we care about privacy:
* append an `iframe` only when needed,
* set its `src` to a blob url (if I guess right it will be Xrayed),
* show hints (labels and rectangles) in this fullscreen, overlay `iframe`,
* show command prompt also in this `iframe` (no change),
* theme file is only present inside the `iframe`.

The biggest downside of showing hints inside the `iframe` is that, we won't be able to color the background of hint rectangles, otherwise we would hide the content under it. However we can show nice rectangles and we can place hint labels more ergonomically if needed.

**Comment by cmcaine** [comment]
Appending an iframe when needed is what we're doing on https://github.com/cmcaine/tridactyl-2-experiments , so, with you on that one ;)

We're not using a blob url for the src, but we could experiment with that. I don't think that we'll cease hinting elements with changes to the content page, though we may put the labels elsewhere. If there's a lot of concern for privacy we might add an option.

Loading theme only to the iframe is something we could do if we change how theming works (which we might do), at present we wouldn't do it because we want users to be able to choose where the iframe itself is placed, etc.

A canny website can still probabilistically profile tridactyl users by noticing that the iframe has been added and realising that not many users get iframes added.

Tridactyl also doesn't work well with privacy.resistfingerprinting in firefox because of firefox bugs. :/

Thanks for your suggestions :)

**Comment by zsugabubus** [comment]
> I don't think that we'll cease hinting elements

Under "ergonomically" I meant not placing multiple labels top of each other, because it makes them unreadable. See for example [wikipedia]( https://en.wikipedia.org/wiki/Death_by_coconut). Anyway, it isn't specific to my `iframe` stuff, I guess bounding rectangles are computed now too.

> we want users to be able to choose where the iframe itself is placed, etc.

I think, I need to clarify this too. I meant a transparent, "fullwindow", overlay `iframe`. Hints and command-line popup are opened inside it, so popup can have border, can be placed anywhere where user wants, can have anything, it just need to be wrapped inside a container element, for example `<div>`.

> A canny website can still probabilistically profile tridactyl users by noticing that the iframe has been added and realising that not many users get iframes added.

Eww... yes... but it just a random `iframe` with an unreadable blob url. Theoretically, ""it can be anything"", but if site guesses "hey, it must be Tridactyl", it couldn't make any differences between Tridactyl users.

> doesn't work well with privacy.resistfingerprinting

I haven't noticed any anomalies.

**Comment by cmcaine** [comment]
Yeah, hint labels aren't placed very well at the moment. It might be a moderately fun thing to fix if you feel like it.

Drawing an iframe across the entire screen means that we must choose to either have the iframe or the window not receive any pointer events. On the retreat we decided that it's useful to have both be able to receive events.

> Eww... it couldn't make any differences between Tridactyl users.

Just knowing that the user adds iframes to pages is quite a few bits of entropy (because very few users do that). Plus the extra bits from e.g. webgl fingerprinting, IP, etc, that is probably enough to uniquely identify the user.

> privacy.resistfingerprinting

There are some reported issues because it messes with keyevents and stuff. Can't remember specifics right now.

We'll take these comments into account when (if?) we get around to rewriting the themes code.

**Comment by bovine3dom** [comment]
We no longer add theme classnames to the root element : )

**Comment by quisquama** [comment]
Is this actually resolved? In the javascript console running `document.getElementsByTagName("html")[0].className` prints `"TridactylThemeDark"`

**Comment by bovine3dom** [comment]
You're right, sorry. I added it back in for compatibility as I didn't want to break any custom themes that were relying on it.

https://github.com/tridactyl/tridactyl/blob/da9bea7edf63dd3ec24bd44c120cb72d6530ee18/src/content/styling.ts#L61

We could add a setting that lets you turn it off.
